### PR TITLE
[Clojure] syntax rework

### PR DESCRIPTION
**Pull request by mitranim**
This PR is a complete rework of the Clojure syntax.

Improvements:

  * The syntax structure is modeled after the AST and the Clojure reader. Should
    handle any formatting, regardless of whitespace and indentation quirks.

  * No unnecessary special cases. Simple and consistent rules. User-defined
    syntax should look and feel the same as standard syntax.

  * All syntactic forms are correctly scoped in file root

  * Much better at handling literals such as numbers, keywords, etc

  * Works for custom defs, e.g. `defroute`

  * Correctly declares `definterface`, `defprotocol`, `deftype`, `defrecord` types

  * Correctly highlights methods in `definterface`, `defprotocol`, `deftype`, `defrecord`

  * Correctly declares functions defined by `defprotocol`

  * Correctly declares names in the presence of metadata, whitespace, comments

  * Correctly handles `defmulti` and `defmethod`

  * Doesn't declare `declare` (noise)

  * Highlights mismatched brackets, helping with balancing

  * Consistent highlighting of function calls

  * ≈ 4 times smaller

  * ≈ 6 times faster

Doesn't detect function parameters and local bindings. I'm not sold on supporting 8+ special assignment forms, each with their own quirks, in a language with macros. It immediately breaks when a user introduces their own variation ([example](https://gist.github.com/Mitranim/148f0a0b51f36ef6a2cd3a0996dc7104) of a slightly different `let` I use in practice). Suggestions are welcome.

Would appreciate review.

**Edit:** while the PR is incubating, the syntax is available as standalone at https://github.com/Mitranim/sublime-clojure

### CODE REVIEW

**Comment by jrappen** [comment]
## 1)

⚠️ There are no tests in your test files...?

* https://www.sublimetext.com/docs/3/syntax.html
* https://www.sublimetext.com/docs/3/syntax.html#testing
* https://www.sublimetext.com/docs/3/scope_naming.html

Compare the test for Clojure:

* [`./Clojure/syntax_test_clojure.clj`](https://github.com/sublimehq/Packages/blob/928a7a6/Clojure/syntax_test_clojure.clj)

Also, your package structure:

* One test file:
  ```text
  /Clojure
      Clojure.sublime-syntax
      syntax_test_clojure.clj
  ```
* several test files:
  ```text
  /Clojure
      Clojure.sublime-syntax
      /tests
          syntax_test_clojure.clj
          syntax_test_clojure_benchmark.clj
          ...
  ```

## 2)

Using whitespace for a vertical separation of blocks is nice to the eyes (at least mine...). For the syntax files as well as the test(s).

Compare the Python subfolder:

[`./Python`](https://github.com/sublimehq/Packages/tree/928a7a6/Python)

also great, LaTeX:

[`./LaTeX`](https://github.com/sublimehq/Packages/tree/928a7a6/LaTeX)

File: Clojure/Clojure.sublime-syntax [hunk 1]
```diff
@@ -1,1123 +1,269 @@
...
+    [-+]?
+    (?:
+      \d+N?                             # int
+      |0[Xx][0-9A-Fa-f]+N?              # hex int
+      |[1-9]\d*[Rr][0-9A-z]+            # n-radix int
+      |\d+/\d+                          # ratio
+      |\d+(?:\.\d+)?(?:[Ee][-+]?\d+)?M? # decimal with scientific notation
+    )
+    (?=[{{non_number_chars}}])
+  reader_macro_scope: constant.other.symbol.reader-macro.clojure
```
Comment (rwols) [inline Clojure/Clojure.sublime-syntax]: You cannot assign variables to scopes. Remove this variable.

File: Clojure/Clojure.sublime-syntax [hunk 2]
```diff
@@ -1,1123 +1,269 @@
...
-              pop: true
-            - include: binding
-    - match: '[^\s]'
-      push:
-        - meta_scope: invalid.illegal.bindings.clojure
-        - match: '[^\]]'
-          pop: true
-  bindings_form:
+    - include: match_expr
+  match_expr:
```
Comment (rwols) [inline Clojure/Clojure.sublime-syntax]: Style: use-dashes-for-contexts, and use_underscores_for_variables.

File: Clojure/Clojure.sublime-syntax [hunk 3]
```diff
@@ -1,1123 +1,269 @@
...
+    - include: match_noise
+    - match: '[)\]}]'
+      scope: invalid.illegal.clojure
+    - match: '\('
```
Comment (rwols) [inline Clojure/Clojure.sublime-syntax]: You can omit the redundant single quotes around this regex, and various other regexes. Download PackageDev and enable the "Sublime Syntax" syntax from that package to better see what is possible and what is not possible within regexes. I highly recommend that package.

File: Clojure/Clojure.sublime-syntax [hunk 4]
```diff
@@ -1,1123 +1,269 @@
...
+      scope: punctuation.section.parens.end.clojure
+      pop: true
+    - match: '\('
+      scope: punctuation.section.parens.begin.clojure
+      push: pop_fn_list_tail
+    - include: match_expr
+  pop_fn_list_tail:
+    - include: match_constant_set_normal_list_tail
+    - match: '{{symbol}}'
+      scope: entity.name.fn.clojure
```
Comment (rwols) [inline Clojure/Clojure.sublime-syntax]: entity.name.function?

File: Clojure/syntax_test_clojure.clj [hunk 5]
```diff
@@ -1,107 +1,666 @@
...
-;;                              ^ invalid.illegal.unexpected-quantifier.regexp
-;;                                       ^^ keyword.operator.quantifier.regexp
-;;                                                 ^ keyword.operator.quantifier.regexp
-
-  (def hello (fn [] #"(?=unclosed_paren"))
-;;                  ^^^^^^^^^^^^^^^^^^^^ string.regexp.clojure
-;;                  ^^ punctuation.definition.string.begin.clojure
-;;                                     ^ punctuation.definition.string.end.clojure
-;;                                      ^ - string.regexp.clojure
+; SYNTAX TEST "Packages/Clojure/Clojure.sublime-syntax"
```
Comment (rwols) [inline Clojure/syntax_test_clojure.clj]: The point of these syntax test files is that you *test* the syntax. For example,
```
"blah \" blah"
; <- punctuation.section
; ^ string.quoted.double
```
Put your build system on "Automatic" and hit "Build" to run these two tests.

**Comment by wbond** [comment]
The existing tests should remain. If scope names need to change, a brief explanation of why should be sufficient.

As @rwols mentioned, it appears your additions to the test file doesn't actually add any assertions. From this I presume you weren't running the Syntax Tests build system when developing this, correct? If so, you should have gotten an indication that there were 0 assertions.

**Comment by Mitranim** [comment]
Hey guys, thanks for the reviews and instructions!

@jrappen Added whitespace for clarity, moved tests under.

@rwols

Thanks for suggesting PackageDev. It's really cool!

Addressed the cosmetics: junk scope variables, quotes, context naming.

Why `entity.name.fn` over `entity.name.function` for lambdas: personally, I prefer keeping lambdas out of the symbol index. Since tail recursion uses `recur`, `fn` names are useful only for better stacktraces, and you never use that name outside the function body. It just interferes with search. Personal preference! Receptive to opinions.

@rwols, @wbond

Tests: I focused on covering as many edge cases as possible, since existing tests were wildly insufficient. Is there a way to automatically add missing assertions to the test file? (If not, this would be really cool to have in PackageDev.)

On a tangential note: what does `Symbol List` do?

**Comment by jrappen** [comment]
@Mitranim

* How are you running the tests? You can run different tests on your syntax files. Open the syntax file, then from the menu choose "Tools > Build with ...". You can see that you can run:
  * Syntax tests
    * what @rwols and me are talking about, see <http://www.sublimetext.com/docs/3/syntax.html#testing>
  * Syntax tests - All Syntaxes
  * Syntax tests - Performance
    * what you are talking about
  * Syntax tests - Regex compatibility
* When looking at [#1259`/files`](https://github.com/sublimehq/Packages/pull/1259/files) it shows me that you removed the syntax file completely. There is a file in your master branch though, not sure what's happening here. I tested on various machines and browers.
* Regarding `Symbol List`: Open a file, then type <kbd>Ctrl</kbd><kbd>R</kbd> or <kbd>Ctrl</kbd><kbd>P</kbd> followed by an <kbd>@</kbd>. For how to use them compare the metadata files in this repo (`*.tmPreferences`) for various syntaxes. There are three available settings:
  * showInSymbolList
  * showInIndexedSymbolList
  * symbolTransformation

**Comment by Mitranim** [comment]
@jrappen 

For now, those tests are for 'manual' reading. I understand how to run the syntax test, but prioritised covering all edge cases first, since it's more useful than half a page of assertions that cover like 5%. Now the question is: is there a tool to automatically add missing assertions to the test? That would be just perfect.

Pretty sure the syntax file is there. Github collapses the diff because it's huge.

Thanks for explaining the symbol lists. Looks like the Clojure symbol list references a scope (`entity.global.clojure`) that's unused in both versions of the syntax. Maybe we should just delete it for now?

**Comment by jrappen** [comment]
@Mitranim I forgot to mention `showInIndexedReferenceList` (Build 3145 or later).

**Comment by wbond** [comment]
@Mitranim No, there is no tool to automatically generate the assertions. From a logical standpoint they wouldn't be testing much since they would just be using what you've already codified as the scope names in the syntax.

When I have some time, I'll go through some of the scope naming. Hopefully you read http://www.sublimetext.com/docs/3/scope_naming.html before writing the syntax.

In terms of `entity.name.fn` and the function list: the default syntaxes aren't a place to customize things to your liking. Plus, the correct way to not have certain scopes indexed or in the symbol list is to create a `.tmPreferences` file in your `Packages/User/` folder that turns off indexing/symbol list for those scopes. We'll definitely want to use the standardized scopes in this syntax, especially since so much work went into it.

Unless there is a good reason to delete the old tests, I'd really prefer they be put back. You should be able to leave them as-is and add your own tests afterwards.

The final question I have for now is: did you run the Syntax Tests: Regex Compatibility build system to see if any incompatible regex features were used?

**Comment by Mitranim** [comment]
Thanks for all the comments folks! Pushed an update following your suggestions, plus extra corrections:

* added autotests
* corrected lambda scopes, banned them from indexing using the "symbol list" file (@jrappen, @wbond, thanks for the hints)
* restored regex support
* added `reify` and `proxy` support
* restored the old test file, but without the meta scope assertions

@wbond  Yep, it passes the regex compliance test.

Please let me know if you feel the giant test suite is too verbose for maintenance. I felt like comprehensiveness was a priority, on the grounds that the previous tests didn't keep the syntax from being buggy. Most of the rules are there because I broke them at some point while writing the syntax, and future changes could do the same.

Kept the "human-only" suite since it's easier to review on a glance. Maybe should remove that.

Haven't added meta scopes yet. I've seen quite a few syntaxes with flat scopes (1-2 levels everywhere), still unsure.

There might be other lurking issues. Please feel free to make suggestions or edits. I'm at the end of my wits here, might need a few days' break. 😅

File: Clojure/tests/syntax_test_clojure.clj [hunk 6]
```diff
@@ -534,7 +534,7 @@
 ;          ^^^^^^ string.clojure
 ;                ^ punctuation.section.parens.end.clojure
 ;                  ^ punctuation.section.parens.begin.clojure
-;                    ^^^^^^
+;                    ^^^^^^ string.clojure
```
Comment (rwols) [inline Clojure/tests/syntax_test_clojure.clj]: maybe use `string.quoted.double.clojure`?

Comment (Mitranim) [reply Clojure/tests/syntax_test_clojure.clj]: renamed `string.clojure` → `string.quoted.double.clojure` everywhere?

**Comment by jrappen** [comment]
* The `uuid` in `*.tmPreferences` files is redundant if not used (... by Python files to access them or whatever).
* While at it, maybe you could fix some of the open clojure issues?

**Comment by Mitranim** [comment]
@jrappen 

Removed unused `uuid`.

Looked into the open Clojure issues:

* #921 looks reasonable; turned out that this syntax uses the stack in a way that makes it difficult to add meta scopes, so this will take some time to figure out; I'll try to look into it later
* #807 is fixed automatically
* #795 is fixed automatically
* #481 is fixed automatically
* not sure if we can do #582, seems like it needs a big buy-in from color scheme developers and users

**Comment by Mitranim** [comment]
Additional improvement: disabled apostrophe auto-pairing outside of comments and strings.

**Comment by Mitranim** [comment]
Corrected detection of n-radix ints. (They no longer accidentally grab non-alphanumeric symbols.)

File: Clojure/tests/syntax_test_clojure.clj [hunk 7]
```diff
@@ -0,0 +1,2007 @@
...
+;                ^- constant
+;                 ^^^^^^^^^^^^^^^ constant.numeric.clojure
+;                                 ^^^^^^^^^^^^^^^ constant.numeric.clojure
+;                                                 ^^^^^^^^^^^^^^^ constant.numeric.clojure
+
+; ## Breaks
+
+  10,20,30
+; ^^ constant.numeric.clojure
+;   ^ comment.punctuation.comma.clojure
```
Comment (wbond) [inline Clojure/tests/syntax_test_clojure.clj]: `comment.punctuation` seems like a funny scope name. Are the commas really comments and not some sort of separator?

File: Clojure/tests/syntax_test_clojure.clj [hunk 8]
```diff
@@ -0,0 +1,2007 @@
...
+; ^^^^ constant.language.clojure
+;     ^ comment.punctuation.comma.clojure
+  true;false;nil
+; ^^^^ constant.language.clojure
+;     ^ comment.line.clojure punctuation.definition.comment
+
+; ## Unaffected
+
+  'nil (true) (nil)
+; ^ constant.other.symbol.reader-macro.clojure
```
Comment (wbond) [inline Clojure/tests/syntax_test_clojure.clj]: Can you explain a little bit about what sort of semantics these `constant.other.symbol.reader-macro` symbols are? It seems like a funny scope name to me, however I am not familiar with Clojure, and only slightly familiar with lisp-style languages.

File: Clojure/tests/syntax_test_clojure.clj [hunk 9]
```diff
@@ -0,0 +1,2007 @@
...
+
+  1234 1234N +1234 +1234N -1234 -1234N
+; ^^^^ constant.numeric.clojure
+;     ^- constant
+;      ^^^^^ constant.numeric.clojure
+;            ^^^^^ constant.numeric.clojure
+;                  ^^^^^^ constant.numeric.clojure
+;                         ^^^^^ constant.numeric.clojure
+;                               ^^^^^^ constant.numeric.clojure
+  0x1234af 0x1234afN 0X1234AF 0X1234AFN
```
Comment (wbond) [inline Clojure/tests/syntax_test_clojure.clj]: What do the various `af` and `n` suffixes mean? We have some conventions we've been applying to recent work on syntaxes where we apply scopes to target storage/type info in the numeric literals. My hunch is they may be applicable here also. See https://github.com/sublimehq/Packages/blob/master/Python/syntax_test_python.py#L935-L940 for an example.

File: Clojure/tests/syntax_test_clojure.clj [hunk 10]
```diff
@@ -0,0 +1,2007 @@
...
+;                                                  ^^^^^^^^^^ constant.numeric.clojure
+  -2r1010 -16r1234af -32r1234az -2R1010 -16R1234AF -32R1234AZ
+; ^^^^^^^ constant.numeric.clojure
+;        ^- constant
+;         ^^^^^^^^^^ constant.numeric.clojure
+;                    ^^^^^^^^^^ constant.numeric.clojure
+;                               ^^^^^^^ constant.numeric.clojure
+;                                       ^^^^^^^^^^ constant.numeric.clojure
+;                                                  ^^^^^^^^^^ constant.numeric.clojure
+  0/10 10/20 30/0
```
Comment (wbond) [inline Clojure/tests/syntax_test_clojure.clj]: What does the `/` represent?

File: Clojure/tests/syntax_test_clojure.clj [hunk 11]
```diff
@@ -0,0 +1,2007 @@
...
+;               ^ punctuation.section.brackets.end.clojure
+;                ^ punctuation.section.parens.end.clojure
+  ([10/20 30/40])
+; ^ punctuation.section.parens.begin.clojure
+;  ^ punctuation.section.brackets.begin.clojure
+;   ^^^^^ constant.numeric.clojure
+;         ^^^^^ constant.numeric.clojure
+;              ^ punctuation.section.brackets.end.clojure
+;               ^ punctuation.section.parens.end.clojure
+  ([100.100 200.200])
```
Comment (wbond) [inline Clojure/tests/syntax_test_clojure.clj]: Are these floats, or do the `.` represent something else?

File: Clojure/tests/syntax_test_clojure.clj [hunk 12]
```diff
@@ -0,0 +1,2007 @@
...
+;                               ^^ constant.language.keyword.clojure
+;                                  ^^ constant.language.keyword.clojure
+;                                     ^^ constant.language.keyword.clojure
+;                                        ^^ constant.language.keyword.clojure
+;                                           ^^ constant.language.keyword.clojure
+  :++ :--
+; ^^^ constant.language.keyword.clojure
+;    ^- constant
+;     ^^^ constant.language.keyword.clojure
+  :blah
```
Comment (wbond) [inline Clojure/tests/syntax_test_clojure.clj]: Is `:blah` really a built in keyword?

File: Clojure/tests/syntax_test_clojure.clj [hunk 13]
```diff
@@ -0,0 +1,2007 @@
...
+  blah:
+  blah::blah
+  /blah
+  blah/
+
+
+
+; # Keywords
+
+  :! :$ :% :& :* :- :_ := :+ :| :< :> :. :/ :?
```
Comment (wbond) [inline Clojure/tests/syntax_test_clojure.clj]: Are these keywords, or more like operators?

File: Clojure/tests/syntax_test_clojure.clj (hunk 7 above)
Comment (Mitranim) [reply Clojure/tests/syntax_test_clojure.clj]: In Clojure/EDN, comments are parsed as whitespace. They disappear from the AST after parsing, so yeah, they're comments.

File: Clojure/tests/syntax_test_clojure.clj (hunk 9 above)
Comment (Mitranim) [reply Clojure/tests/syntax_test_clojure.clj]: `0x_______` is a hex integer literal, which may include `[A-Fa-f]`.

File: Clojure/tests/syntax_test_clojure.clj (hunk 10 above)
Comment (Mitranim) [reply Clojure/tests/syntax_test_clojure.clj]: In Clojure, `N/M`, where N and M are integers, is a ratio literal. Ratios are a special numeric data type.

File: Clojure/tests/syntax_test_clojure.clj (hunk 11 above)
Comment (Mitranim) [reply Clojure/tests/syntax_test_clojure.clj]: These are floats. Like many other tests, this one looks funny because it tests for funny mistakes I made when developing the syntax.

File: Clojure/tests/syntax_test_clojure.clj (hunk 12 above)
Comment (Mitranim) [reply Clojure/tests/syntax_test_clojure.clj]: Clojure doesn't have any built-in syntactic keywords as seen in other languages, since it's defined in terms of EDN data structures. The keyword data type is the same as Erlang or Ruby atoms. They're literals similar to strings, interned for performance and used as map keys.

File: Clojure/tests/syntax_test_clojure.clj (hunk 13 above)
Comment (Mitranim) [reply Clojure/tests/syntax_test_clojure.clj]: There are actual valid keywords. On a side note, Clojure doesn't have any operators; things like `+` and `=` are regular identifiers.

File: Clojure/tests/syntax_test_clojure.clj (hunk 8 above)
Comment (Mitranim) [reply Clojure/tests/syntax_test_clojure.clj]: Come to think of, this scope name is probably wrong. Reader macros are basically unary prefix operators that modify the next piece of AST as it's being read. I didn't want to mark them as regular operators since they work at parse time, not runtime, and it's nice if the color scheme can make a distinction, hence the current name. In many syntaxes, `constant.other.symbol` receives special highlighting. But we should probably change this.

**Comment by Mitranim** [comment]
One issue that I see with the current implementation is that it's written in a way that makes it difficult to add meta scopes. Can address this if the need arises, but it just felt like a low priority.

File: Clojure/tests/syntax_test_clojure.clj (hunk 9 above)
Comment (Mitranim) [reply Clojure/tests/syntax_test_clojure.clj]: Ooh I see about storage. By that logic, the `N` at the end should be `storage.type.numeric.bigint.clojure`. I think it should still be part of the `constant.numeric` scope, since it's part of the number literal, so it needs a double scope.

Comment (FichteFoll) [reply Clojure/tests/syntax_test_clojure.clj]: Yes, that's also how it is in Python (and other syntaxes). Also note the `punctuation.definition.numeric.hexadecimal.python` scopes.

File: Clojure/tests/syntax_test_clojure.clj (hunk 10 above)
Comment (FichteFoll) [reply Clojure/tests/syntax_test_clojure.clj]: How about adding a `punctuation.separator.ratio-literal.clojure` or similar scope to the slash?

File: Clojure/tests/syntax_test_clojure.clj (hunk 13 above)
Comment (FichteFoll) [reply Clojure/tests/syntax_test_clojure.clj]: There is a `keyword` scope. Why did you not choose a descendant of that?

File: Clojure/tests/syntax_test_clojure.clj (hunk 12 above)
Comment (wbond) [reply Clojure/tests/syntax_test_clojure.clj]: In that case they should be `constant.other`, not `constant.language`: https://www.sublimetext.com/docs/3/scope_naming.html#constant.

File: Clojure/tests/syntax_test_clojure.clj (hunk 13 above)
Comment (wbond) [reply Clojure/tests/syntax_test_clojure.clj]: So it would see to me from your description that the semantics of `:+` and `:blah` are the same, right? They should either be scoped the same, or if by convention they are used as "operators" we could scope them as such. I don't really see a reason they should be keywords, however.

Comment (FichteFoll) [reply Clojure/tests/syntax_test_clojure.clj]: Currently they both have the `constant.language.keyword.clojure` scope.

I don't see any tests against a punctuation scope here, so scoping the colon as `punctuation.definition.keyword.clojure` sounds like a good addition.

Alternatively, if these are more similar to Ruby's symbol, maybe draw some inspiration for scope names from that?

Comment (Mitranim) [reply Clojure/tests/syntax_test_clojure.clj]: AFAICT, Ruby symbols are exactly like Clojure keywords. Looks like we all agree to follow the example and use `constant.other.keyword.clojure` for the entire constant and `punctuation.definition.keyword.clojure` for the colon.

File: Clojure/tests/syntax_test_clojure.clj (hunk 10 above)
Comment (Mitranim) [reply Clojure/tests/syntax_test_clojure.clj]: Sure. `punctuation.separator.ratio.clojure` sounds good to me.

File: Clojure/tests/syntax_test_clojure.clj (hunk 8 above)
Comment (Mitranim) [reply Clojure/tests/syntax_test_clojure.clj]: Not sure if `keyword.macro.clojure` or `keyword.operator.macro.clojure` makes sense? Looks like these fall under `keyword` variety, and I want a scope that's different from `keyword.operator` but not too Clojure-specific. What would you suggest?

**Comment by Mitranim** [comment]
Addressed the comments above.

**Comment by Mitranim** [comment]
Pretty satisfied with how it works. Not strictly "done" (see the above comments on meta scopes), but haven't felt the need for more changes in the past few months. Might be stable enough to move forward with the PR for the sake of other Sublime/Clojure users, all 3 of them (?).

**Comment by wbond** [comment]
I am going to hold off merging any significant work right now since the dev cycle is almost complete. We'll get this merged in at the beginning of the next dev cycle.

My only concern, is based on comments, it sounds as if there may be somewhat of a regression around function definitions. Is it just parameters that aren't highlighted, or the whole function definitions themselves? I think it is worth extra effort to support Goto Definition if this currently doesn't track the definitions of functions. Please excuse my lack of memory in regards to this PR.

**Comment by Mitranim** [comment]
Yeah I just don't know the process, was wondering.

Symbol indexing works perfectly. I just banned inline lambdas from being indexed, since you never want to "goto" them anyway.

File: Clojure/Clojure.sublime-settings [hunk 14]
```diff
@@ -0,0 +1,7 @@
+{
+  "word_separators": "/\\()[]{}\"'`,:;@#^~.",
+  "smart_indent": false,
+  "detect_indentation": false,
+  "tab_size": 1,
```
Comment (wbond) [inline Clojure/Clojure.sublime-settings]: This seems funny to me. Why would the tab size be hardcoded to 1 for all Clojure files?

File: Clojure/Clojure.sublime-settings [hunk 15]
```diff
@@ -0,0 +1,7 @@
...
+  "draw_indent_guides": false
```
Comment (wbond) [inline Clojure/Clojure.sublime-settings]: Is this related to the tab size of 1?

File: Clojure/Default.sublime-keymap [hunk 16]
```diff
@@ -0,0 +1,26 @@
+[
```
Comment (wbond) [inline Clojure/Default.sublime-keymap]: This file should use tabs for indentation

File: Clojure/Clojure.sublime-settings (hunk 14 above)
Comment (Mitranim) [reply Clojure/Clojure.sublime-settings]: Most Clojure code follows the same indentation convention, which sometimes involves 1-space indents. The [Lispindent](https://packagecontrol.io/packages/lispindent) package automates this style, but when refactoring, you often need to shift a block by 1 space, and it's needlessly painful if the indentation is not set to 1.

Example from the test file: [[1]](https://github.com/Mitranim/Packages/blob/c035c0a72b273069eba8dafb0962c8f0434e8fd9/Clojure/tests/syntax_test_clojure.clj#L2050-L2051).

File: Clojure/Clojure.sublime-settings (hunk 15 above)
Comment (Mitranim) [reply Clojure/Clojure.sublime-settings]: Yeah. There's a reason for these settings.

* `smart_indent` doesn't play well with Lisp code
* `detect_indentation` can get confused by traditional Clojure indentation, and there's very rarely any reason for indent other than 1
* `draw_indent_guides` creates needless noise with indent 1

File: Clojure/Default.sublime-keymap (hunk 16 above)
Comment (Mitranim) [reply Clojure/Default.sublime-keymap]: Fixed.

File: Clojure/Clojure.sublime-settings (hunk 14 above)
Comment (Mitranim) [reply Clojure/Clojure.sublime-settings]: The most common reason for indent 1 is when writing nested data structures, such as Hiccup-style markup:

```clj
(def page []
  [:div.class0
   [:div.class1
    [:div.class2
     ...]]])
```

Comment (FichteFoll) [reply Clojure/Clojure.sublime-settings]: If most code is formatted that way, the automatic detection would be able to catch this, wouldn't it? I'd be very vary of overriding indentation settings for all users, unless the format basically requires you to use it (kind of like YAML).

**Comment by FichteFoll** [comment]
If most code is formatted that way, the automatic detection would be able to catch this, wouldn't it? I'd be very wary of overriding indentation settings for all users, unless the format basically requires you to use it (kind of like YAML).

However, if you believe in good faith that no sane user should ever use any other settings than the ones you provided (because they are very troublesome to work with or outright don't work at all), I certainly do see reasons for having them.

**Comment by Mitranim** [comment]
Automatic indent detection is based on the first indented lines. Most Clojure files start with this:

```clj
(ns my-domain.my-namespace
  "Namespace docstring"
  (:require
    [namespace0]
    [namespace1])
  (:import
    [class0]
    [class1]))
```

Conventionally, the next line inside `()` is intended by 2 spaces, whereas the next line inside `[]` or `{}` is indented by 1 space. Automatic detection would always pick 2 spaces, which would be incorrect.

The main reason why I feel 1 space is more convenient is refactoring. Sometimes you need to shift a block of code by an uneven number of spaces due to the aforementioned conventions. If the block doesn't have any newlines in it, multi-cursor editing works fine. But if the block does have newlines and you're shifting to the left, you can't just multi-cursor: it would delete a newline and then delete something extra.

This is just my experience. If someone finds this default inconvenient, they're welcome to complain and we'll change it.

**Comment by wbond** [comment]
I'm keen on getting this in so that it gets testing from other Clojure users during the next dev cycle. That said, I'm not convinced about the indentation settings tweak. Could you remove those and propose them in another PR, and hopefully we can get some feedback from other ST Clojure users in the future?

**Comment by Mitranim** [comment]
Glad to hear this is getting progress.

No problem with removing the tweak, but are you sure going 2 → 1 is the right order? I assume that for most users, if 2 is inconvenient, they won't provide feedback saying 1 is better; nobody will even try 1 because it's so radical. Going 1 → 2, if folks find 1 inconvenient, they're somewhat more likely to complain, and we're more likely to receive feedback.

**Comment by ItamarShDev** [comment]
I will be glad to help feedback-wise.
I use clojure daily at work :)

**Comment by rwols** [comment]
@ItamarShDev have you tried this diff on your codebase?
